src/base/misc/disks.c

```diff
@@ -532,10 +533,12 @@ static void image_auto(struct disk *dp)
           dp->dev_name);
     leavedos(20);
   }
-  dp->num_secs = (unsigned long long)dp->tracks * dp->heads * dp->sectors;
+  filesize = lseek64(dp->fdesc, 0, SEEK_END);
+  dp->num_secs = (filesize + SECTOR_SIZE - 1) / SECTOR_SIZE;
```

Comment:
I don't understand what does this mean.
I think this patch should be merged with
another one or removed, as it doesn't
seem to represent a sensible change.

Reply:
Note I moved this into the conditional's non-magic-header case in the next commit. It sets the image's num_secs from the file size. I don't know how else to explain this?

Reply:
But I don't think you can explain this:
`-  dp->num_secs = (unsigned long long)dp->tracks * dp->heads * dp->sectors;`
This equation should always be honored,
so I don't understand how it can be removed.
I think we should re-calculate "tracks" first.
If the image size does not match the
XX/255/63 layout, then we should just
truncate it to the nearest match.

Reply:
In the next commit I did restore the tracks * heads * sectors calculation for the headered case, because I expect that the header spells out how large the file should be. However, for the non-header autosensing, I made it so that 255 heads and 63 sectors are assumed, while the dp->num_secs and subsequently dp->tracks are calculated from the image file's size.

Also, I think the file accesses shouldn't be *truncated* to the nearest multiple of 255 heads * 63 sectors, but rather the CHS geometry should be rounded up so that all the sectors are accessible, including (if applicable) ones in a partial last cylinder.

Reply:
> In the next commit I did restore the tracks * heads * sectors calculation for the headered case

But that's why I asked you to merge it
with another commit so that together they
start making sense. We should try to avoid
partial commits because things should be
bisectable, plus its difficult to review. To merge
commits you need to use `git rebase -i`.

> Also, I think the file accesses shouldn't be
> truncated to the nearest multiple of 255 heads * 63
> sectors, but rather the CHS geometry should be rounded up

But why do you want it that way?
We try to model real bios, not being creative. :)

Reply:
> But that's why I asked you to merge it
with another commit so that together they
start making sense. We should try to avoid
partial commits because things should be
bisectable, plus its difficult to review.

This would still be bisectable in this case, I think. However, I will rebase and push again.

> But why do you want it that way?

Because I expect that the entire image file should be accessible (using LBA or, if in range, CHS), not just the image truncated at some essentially arbitrary limit. I don't know that any BIOS really truncates this way, either.

Reply:
> using LBA or, if in range, CHS

Maybe if only for LBA and in-range CHS, then yes.
But you also do
```
+    dp->tracks = (dp->num_secs + (dp->heads * dp->sectors - 1))
+		  / (dp->heads * dp->sectors);
```
which means you round-up the CHS geo
either, which will likely mean some sector-not-found
stuff at the end of an image. I suppose mkfatimage16
should create "CHS-aligned" images.

Reply:
This tracks calculation isn't used by mkfatimage16 images at all (unless made raw by using -r), but I indeed intended that for raw images it should have "sector not found" within the partial cylinder, if the image file isn't aligned to the geometry.